File: android/MultiAIIM/app/src/test/java/com/multiaicode/remoteim/ChatStateTest.java
package com.multiaicode.remoteim;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

import java.util.List;

public class ChatStateTest {
    @Test
    public void upsertsContactAndQueuesOutgoingText() {
        ChatState state = new ChatState("android-user");

        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");
        RemoteIMMessage message = state.queueOutgoingText("检查构建");

        assertEquals("mac-office", state.selectedPeerId());
        assertEquals(1, state.contacts().size());
        assertEquals("检查构建", message.text());
        assertEquals(RemoteIMMessage.Direction.OUTGOING, message.direction());
        assertEquals(RemoteIMMessage.Status.PENDING, message.status());
        assertEquals(List.of(message), state.messagesWith("mac-office"));
    }

    @Test
    public void recordsIncomingTextAndAutoAddsContact() {
        ChatState state = new ChatState("android-user");

        RemoteIMMessage message = state.receiveText("处理完成", "mac-office");

        assertEquals("处理完成", message.text());
        assertEquals(RemoteIMMessage.Direction.INCOMING, message.direction());
        assertEquals(RemoteIMMessage.Status.RECEIVED, message.status());
        assertEquals("mac-office", state.contacts().get(0).userId());
    }

    @Test
    public void queuesOutgoingImageWithAttachment() {
        ChatState state = new ChatState("android-user");
        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");

        RemoteIMMessage message = state.queueOutgoingImage("/tmp/photo.png", 640, 480, 4096);

        assertEquals("[图片消息] photo.png", message.text());
        assertNotNull(message.imageAttachment());
        assertEquals("/tmp/photo.png", message.imageAttachment().localPath());
        assertEquals(640, message.imageAttachment().width());
        assertEquals(480, message.imageAttachment().height());
        assertEquals(4096, message.imageAttachment().sizeBytes());
    }

    @Test
    public void queuesAndReceivesVoiceMessages() {
        ChatState state = new ChatState("android-user");
        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");

        RemoteIMMessage outgoing = state.queueOutgoingVoice("/tmp/out.m4a", 3);
        RemoteIMMessage incoming = state.receiveVoice("/tmp/in.m4a", 4, "mac-office");

        assertEquals("[语音消息 3s]", outgoing.text());
        assertEquals(RemoteIMMessage.Direction.OUTGOING, outgoing.direction());
        assertNotNull(outgoing.voiceAttachment());
        assertEquals("[语音消息 4s]", incoming.text());
        assertEquals(RemoteIMMessage.Direction.INCOMING, incoming.direction());
        assertNotNull(incoming.voiceAttachment());
    }

    @Test
    public void messageStatusCanBeUpdated() {
        ChatState state = new ChatState("android-user");
        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");
        RemoteIMMessage message = state.queueOutgoingText("ping");

        assertTrue(state.updateMessageStatus(message.id(), RemoteIMMessage.Status.SENT));

        assertEquals(RemoteIMMessage.Status.SENT, state.messagesWith("mac-office").get(0).status());
        assertFalse(state.updateMessageStatus("missing", RemoteIMMessage.Status.SENT));
    }
}

File: desktop/qt-im/tests/ChatStateTest.cpp
#include <QtTest/QtTest>

#include "model/ChatState.h"

class ChatStateTest : public QObject {
    Q_OBJECT

private slots:
    void queuesOutgoingText();
    void removesAicliProtocolPrefixFromIncomingText();
    void removesAicliProtocolPrefixFromRestoredIncomingText();
    void receivesIncomingImage();
    void receivesIncomingFile();
    void updatesMessageStatus();
    void returnsPeerMessagesChronologically();
    void removesContactAndMessages();
};

void ChatStateTest::queuesOutgoingText() {
    ChatState state("desktop-user");
    state.upsertContact(RemoteIMContact{"ios-user", "ios-user"});
    state.selectPeer("ios-user");

    const RemoteIMMessage message = state.queueOutgoingText("ping");

    QCOMPARE(message.fromUserId, QString("desktop-user"));
    QCOMPARE(message.toUserId, QString("ios-user"));
    QCOMPARE(message.text, QString("ping"));
    QCOMPARE(message.direction, RemoteIMMessageDirection::Outgoing);
    QCOMPARE(message.status, RemoteIMMessageStatus::Pending);
    QCOMPARE(state.messagesWith("ios-user").size(), 1);
}

void ChatStateTest::removesAicliProtocolPrefixFromIncomingText() {
    ChatState state("desktop-user");
    const QString hiddenPrefix = QStringLiteral("\u2063\u200B\u200C\u200D\u2063");

    const RemoteIMMessage hiddenMarked = state.receiveText(
        "electron-user",
        hiddenPrefix + QStringLiteral("## 结论\n\nMarkdown 正文")
    );
    const RemoteIMMessage legacyMarked = state.receiveText(
        "electron-user",
        QStringLiteral("【AICLI 输出】\n## 结果")
    );

    QCOMPARE(hiddenMarked.text, QStringLiteral("## 结论\n\nMarkdown 正文"));
    QCOMPARE(legacyMarked.text, QStringLiteral("## 结果"));
}

void ChatStateTest::removesAicliProtocolPrefixFromRestoredIncomingText() {
    ChatState state("desktop-user");
    const QString hiddenPrefix = QStringLiteral("\u2063\u200B\u200C\u200D\u2063");
    RemoteIMMessage message;
    message.fromUserId = QStringLiteral("electron-user");
    message.toUserId = QStringLiteral("desktop-user");
    message.text = hiddenPrefix + QStringLiteral("# Win/Mac 每周 Crash 详细报表");
    message.direction = RemoteIMMessageDirection::Incoming;

    state.appendMessageForRestore(message);

    QCOMPARE(state.messagesWith(QStringLiteral("electron-user")).first().text,
             QStringLiteral("# Win/Mac 每周 Crash 详细报表"));
}

void ChatStateTest::receivesIncomingImage() {
    ChatState state("desktop-user");

    const RemoteIMMessage message = state.receiveImage("ios-user", "/tmp/a.jpg", 640, 480, 1024);

    QCOMPARE(message.fromUserId, QString("ios-user"));
    QCOMPARE(message.toUserId, QString("desktop-user"));
    QCOMPARE(message.text, QString("[图片消息] a.jpg"));
    QCOMPARE(message.direction, RemoteIMMessageDirection::Incoming);
    QCOMPARE(message.status, RemoteIMMessageStatus::Received);
    QVERIFY(message.hasImage);
    QCOMPARE(state.contacts().size(), 1);
    QCOMPARE(state.selectedPeerId(), QString("ios-user"));
}

void ChatStateTest::receivesIncomingFile() {
    ChatState state("desktop-user");

    const RemoteIMMessage message = state.receiveFile(
        "ios-user",
        "/tmp/report.md",
        "report.md",
        "text/markdown",
        4096
    );

    QCOMPARE(message.fromUserId, QString("ios-user"));
    QCOMPARE(message.toUserId, QString("desktop-user"));
    QCOMPARE(message.text, QString("[文件消息] report.md"));
    QCOMPARE(message.direction, RemoteIMMessageDirection::Incoming);
    QCOMPARE(message.status, RemoteIMMessageStatus::Received);
    QVERIFY(message.hasFile);
    QCOMPARE(message.file.localPath, QString("/tmp/report.md"));
    QCOMPARE(message.file.fileName, QString("report.md"));
    QCOMPARE(message.file.mimeType, QString("text/markdown"));
    QCOMPARE(message.file.sizeBytes, qint64(4096));
}

void ChatStateTest::updatesMessageStatus() {
    ChatState state("desktop-user");
    state.upsertContact(RemoteIMContact{"ios-user", "ios-user"});
    state.selectPeer("ios-user");
    const RemoteIMMessage message = state.queueOutgoingText("ping");

    QVERIFY(state.updateMessageStatus(message.id, RemoteIMMessageStatus::Sent));

    QCOMPARE(state.messagesWith("ios-user").first().status, RemoteIMMessageStatus::Sent);
}

void ChatStateTest::returnsPeerMessagesChronologically() {
    ChatState state("desktop-user");

    RemoteIMMessage newest;
    newest.id = QStringLiteral("newest");
    newest.fromUserId = QStringLiteral("ios-user");
    newest.toUserId = QStringLiteral("desktop-user");
    newest.text = QStringLiteral("30s");
    newest.createdAtMillis = 30'000;

    RemoteIMMessage oldest = newest;
    oldest.id = QStringLiteral("oldest");
    oldest.text = QStringLiteral("10s");
    oldest.createdAtMillis = 10'000;

    RemoteIMMessage middle = newest;
    middle.id = QStringLiteral("middle");
    middle.text = QStringLiteral("20s");
    middle.createdAtMillis = 20'000;

    state.appendMessageForRestore(newest);
    state.appendMessageForRestore(oldest);
    state.appendMessageForRestore(middle);

    const QList<RemoteIMMessage> messages = state.messagesWith(QStringLiteral("ios-user"));

    QCOMPARE(messages.size(), 3);
    QCOMPARE(messages.at(0).text, QStringLiteral("10s"));
    QCOMPARE(messages.at(1).text, QStringLiteral("20s"));
    QCOMPARE(messages.at(2).text, QStringLiteral("30s"));
}

void ChatStateTest::removesContactAndMessages() {
    ChatState state("desktop-user");
    state.upsertContact(RemoteIMContact{"ios-user", "iPhone"});
    state.upsertContact(RemoteIMContact{"other-user", "Other"});
    state.selectPeer("ios-user");
    state.receiveText("ios-user", "hello");
    state.receiveText("other-user", "keep");

    state.removeContactAndMessages(" ios-user ");

    QCOMPARE(state.contacts().size(), 1);
    QCOMPARE(state.contacts().first().userId, QString("other-user"));
    QCOMPARE(state.messagesWith("ios-user").size(), 0);
    QCOMPARE(state.messagesWith("other-user").size(), 1);
    QCOMPARE(state.selectedPeerId(), QString("other-user"));
}

QTEST_MAIN(ChatStateTest)
#include "ChatStateTest.moc"
